lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1144 @@
+    ///   shutdown both halves of the shared TCP stream. The Shutdown command
+    ///   ensures that the Reader / Writer will unblock if they are currently
+    ///   blocked on `recv()`. The TCP stream shutdown ensures that they will
+    ///   unblock if they are currently blocked on `read()` or `write()`
+    ///   respectively.
+    ///
+    /// In cases (1) and (3), the disconnect was NOT initiated by
+    /// Rust-Lightning, so the Reader / Writer notify the PeerManager using
+    /// `socket_disconnected()`.
+    fn disconnect_socket(&mut self) {
+        let _ = self.reader_cmd_tx.try_send(ReaderCommand::Shutdown);
+        let _ = self.writer_cmd_tx.try_send(WriterCommand::Shutdown);
+        let _ = self.tcp_disconnector.shutdown();
+    }
+}
+
+/// An actor that synchronously handles the read() events emitted by the socket.
+struct Reader<CMH, RMH, L, UMH>
+where
+    CMH: ChannelMessageHandler + 'static + Send + Sync,
+    RMH: RoutingMessageHandler + 'static + Send + Sync,
+    L: Logger + 'static + ?Sized + Send + Sync,
+    UMH: CustomMessageHandler + 'static + Send + Sync,
+{
+    inner: TcpReader,
+    peer_manager: Arc<PeerManager<SyncSocketDescriptor, Arc<CMH>, Arc<RMH>, Arc<L>, Arc<UMH>>>,
+    descriptor: SyncSocketDescriptor,
+    reader_cmd_rx: Receiver<ReaderCommand>,
+    read_paused: bool,
+}
+impl<CMH, RMH, L, UMH> Reader<CMH, RMH, L, UMH>
+where
+    CMH: ChannelMessageHandler + 'static + Send + Sync,
+    RMH: RoutingMessageHandler + 'static + Send + Sync,
+    L: Logger + 'static + ?Sized + Send + Sync,
+    UMH: CustomMessageHandler + 'static + Send + Sync,
+{
+    fn new(
+        reader: TcpReader,
+        peer_manager: Arc<PeerManager<SyncSocketDescriptor, Arc<CMH>, Arc<RMH>, Arc<L>, Arc<UMH>>>,
+        descriptor: SyncSocketDescriptor,
+        reader_cmd_rx: Receiver<ReaderCommand>,
+    ) -> Self {
+        Self {
+            inner: reader,
+            peer_manager,
+            descriptor,
+            reader_cmd_rx,
+            read_paused: false,
+        }
+    }
+
+    /// Handle read events, or wait for the next `ReaderCommand` if reads are
+    /// paused. This implementation avoids busy loops and lets the thread go to
+    /// sleep whenever reads or channel commands are pending.
+    ///
+    /// - If `self.read_paused`, block on `self.reader_cmd_rx.recv()` and handle
+    ///   any commands accordingly.
+    /// - If `!self.read_paused`, block on `self.inner.read()` and handle any
+    ///   read events accordingly.
+    /// - In between each event, do a non-blocking check for `ReaderCommand`s.
+    fn run(&mut self) {
+        // 8KB is nice and big but also should never cause any issues with stack
+        // overflowing.
+        let mut buf = [0; 8192];
+
+        loop {
+            // Every time this line is reached, read_paused == false.
+            // Do a non-blocking try_recv() to check for commands
+            if self.do_try_recv() {
```

Comment:
kind of confusing to read. could do like

```rust
let is_shutdown = self.do_try_recv();
if is_shutdown {
  break;
}
```

Reply:
same for below do_recv, do_read

Reply:
hmm, could also do like

```rust
enum ReaderState {
  Ready,
  CheckChannel,
  Paused,
  Shutdown,
}
```

then like

```rust
loop {
  match self.state {
    Shutdown => break,
    Paused => self.do_recv(),
    CheckChannel => self.do_try_recv(),
    Ready => self.do_read(&mut buf),
  }
}
```

Reply:
Cool. Had that before but thought you didn't like it

Reply:
ooo that's really clean. Will implement after guan dan loe